src/ert/config/forward_model_step.py

```diff
@@ -1,25 +1,126 @@
 import logging
-import os
-import os.path
 from dataclasses import dataclass, field
-from typing import Dict, List, Optional, Tuple, no_type_check
+from typing import (
+    Dict,
+    List,
+    Optional,
+    TypedDict,
+)
+
+from typing_extensions import Unpack
 
 from ert.substitution_list import SubstitutionList
 
-from .parse_arg_types_list import parse_arg_types_list
 from .parsing import (
-    ConfigValidationError,
-    ForwardModelStepKeys,
     SchemaItemType,
-    init_forward_model_schema,
-    lark_parse,
 )
 
 logger = logging.getLogger(__name__)
 
 
+class ForwardModelInvalidCallError(Exception):
+    """Thrown when the user calls the forward model incorrectly.
+
+    Can be subtyped by the implementation of ForwardModelStepPlugin and
+    thrown from `validate_pre_realization_run` or `validate_pre_experiment`.
+    """
+
+
+class ForwardModelStepJSON(TypedDict):
+    """
+    A dictionary containing information about how a forward model step should be run
+    on the queue.
+
+    Attributes:
+        executable: The name of the executable to be run
+        target_file: Name of file expected to be produced after the forward
+            model step is executed successfully.
+            This file is used for the legacy ERT queue driver, and may be deprecated.
+        error_file: Name of file expected to be produced after a forward
+            model step errors.
+            This file is used for the legacy ERT queue driver, and may be deprecated.
+        start_file: Name of file expected to be produced upon start of
+            a forward model step's execution.
+            This file is used for the legacy ERT queue driver, and may be deprecated.
+        stdout: File where this forward model step's stdout is written
+        stderr: File where this forward model step's stderr is written
+        stdin: File where this forward model step's stdin is written
+        argList: List of command line arguments to be given to the executable.
+        environment: Dictionary of environment variables to inject into the
+            environment of the forward model step run
+        exec_env: Dictionary of environment variables to inject into the execution
+            environment of RMS
+        max_running_minutes: Maximum runtime in minutes. If the forward model step
+            takes longer than this, the job is requested to be cancelled.
+    """
+
+    name: str
+    executable: List[str]
+    target_file: str
+    error_file: str
+    start_file: str
+    stdout: str
+    stderr: str
+    stdin: str
+    argList: List[str]
+    environment: Dict[str, str]
+    exec_env: Dict[str, str]
+    max_running_minutes: int
+
+
+class ForwardModelStepOptions(TypedDict, total=False):
+    stdin_file: Optional[str]
+    stdout_file: Optional[str]
+    stderr_file: Optional[str]
+    start_file: Optional[str]
+    target_file: Optional[str]
+    error_file: Optional[str]
```

Comment:
With TypedDict you need to use `NotRequired to indicate that the dictionary does not need to contain the keyword: https://typing.readthedocs.io/en/latest/spec/callables.html#required-and-non-required-keys .